Impossible to import RolePolicy with newest admin-client against Keycloak 24

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

admin/client-java

### Describe the bug

When using admin client of newest version against Keycloak 24, there can be issues when creating role policy. Running this:

```
RolePolicyRepresentation rolePolicy = new RolePolicyRepresentation();

rolePolicy.setName("Only User Policy");
rolePolicy.addRole("user");

clientResource.authorization().policies().role().create(rolePolicy).close();
```

fails on the server side with the exception:

```
2024-08-13 10:52:29,683 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 2024-08-13 10:52:29,681 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-2) Uncaught server error: java.lang.RuntimeException: Failed to deserialize JSON using policy provider for type [role].
2024-08-13 10:52:29,683 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at org.keycloak.authorization.admin.PolicyTypeService.doCreateRepresentation(PolicyTypeService.java:79)
2024-08-13 10:52:29,684 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at org.keycloak.authorization.admin.PolicyService.create(PolicyService.java:119)
2024-08-13 10:52:29,684 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at org.keycloak.authorization.admin.PolicyService$quarkusrestinvoker$create_195cd61c18e9138a5e834481c705d4fd435d9cb0.invoke(Unknown Source)
2024-08-13 10:52:29,684 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
2024-08-13 10:52:29,684 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
2024-08-13 10:52:29,684 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
2024-08-13 10:52:29,684 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
2024-08-13 10:52:29,684 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
2024-08-13 10:52:29,685 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
2024-08-13 10:52:29,685 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
2024-08-13 10:52:29,685 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
2024-08-13 10:52:29,685 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
2024-08-13 10:52:29,685 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at java.base/java.lang.Thread.run(Thread.java:840)
2024-08-13 10:52:29,685 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: Caused by: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "fetchRoles" (class org.keycloak.representations.idm.authorization.RolePolicyRepresentation), not marked as ignorable (13 known properties: "resources", "scopesData", "name", "decisionStrategy", "policies", "owner", "type", "id", "description", "roles", "logic", "scopes", "resourcesData"])
2024-08-13 10:52:29,685 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT:  at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 151] (through reference chain: org.keycloak.representations.idm.authorization.RolePolicyRepresentation["fetchRoles"])
2024-08-13 10:52:29,685 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException.from(UnrecognizedPropertyException.java:61)
2024-08-13 10:52:29,686 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at com.fasterxml.jackson.databind.DeserializationContext.handleUnknownProperty(DeserializationContext.java:1153)
2024-08-13 10:52:29,686 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at com.fasterxml.jackson.databind.deser.std.StdDeserializer.handleUnknownProperty(StdDeserializer.java:2224)
2024-08-13 10:52:29,686 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownProperty(BeanDeserializerBase.java:1793)
2024-08-13 10:52:29,686 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownVanilla(BeanDeserializerBase.java:1771)
2024-08-13 10:52:29,686 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:316)
2024-08-13 10:52:29,686 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
2024-08-13 10:52:29,686 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:342)
2024-08-13 10:52:29,686 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4899)
2024-08-13 10:52:29,687 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3846)
2024-08-13 10:52:29,687 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3814)
2024-08-13 10:52:29,687 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at org.keycloak.util.JsonSerialization.readValue(JsonSerialization.java:75)
2024-08-13 10:52:29,687 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	at org.keycloak.authorization.admin.PolicyTypeService.doCreateRepresentation(PolicyTypeService.java:77)
2024-08-13 10:52:29,687 INFO  [tc.qua.io/.0] (docker-java-stream-693675543) STDOUT: 	... 12 more
```

## Cause

The field `fetchRoles` is currently always sent by the admin client due the fact that it is marked on the class `RolePolicyRepresentation` as primitive `boolean` . Updating to `Boolean` similarly like all other representation objects are using (See for example `RealmRepresentation`) makes it possible that admin-client does not include the field (if not set explicitly) and hence makes it possible to create the role in Keycloak 24 where the field `fetchRoles` does not yet exists on the `RolePolicyRepresentation`.


### Version

admin-client nightly with Keycloak server 24.0.5

### Regression

- [ ] The issue is a regression

### Expected behavior

It is possible to create rolePolicy with newest admin-client to Keycloak 24.0.5 (if the field `fetchRoles` is omitted)


### Actual behavior

It is possible to create rolePolicy with newest admin-client to Keycloak 24.0.5 (if the field `fetchRoles` is omitted)

### How to Reproduce?

See the description

### Anything else?

Related PR to keycloak-client: https://github.com/keycloak/keycloak-client/pull/22

--- a/authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/role/RolePolicyProvider.java
+++ b/authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/role/RolePolicyProvider.java
@@ -62,9 +62,10 @@ public class RolePolicyProvider implements PolicyProvider {
             RoleModel role = realm.getRoleById(roleDefinition.getId());
 
             if (role != null) {
-                boolean hasRole = hasRole(identity, role, realm, authorizationProvider, policyRep.isFetchRoles());
+                boolean isFetchRoles = policyRep.isFetchRoles() != null && policyRep.isFetchRoles();
+                boolean hasRole = hasRole(identity, role, realm, authorizationProvider, isFetchRoles);
 
-                if (!hasRole && roleDefinition.isRequired()) {
+                if (!hasRole && roleDefinition.isRequired() != null && roleDefinition.isRequired()) {
                     evaluation.deny();
                     return;
                 } else if (hasRole) {

--- a/authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/role/RolePolicyProviderFactory.java
+++ b/authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/role/RolePolicyProviderFactory.java
@@ -142,7 +142,9 @@ public class RolePolicyProviderFactory implements PolicyProviderFactory<RolePoli
     }
 
     private void updateRoles(Policy policy, RolePolicyRepresentation representation, AuthorizationProvider authorization) {
-        policy.putConfig("fetchRoles", String.valueOf(representation.isFetchRoles()));
+        if (representation.isFetchRoles() != null) {
+            policy.putConfig("fetchRoles", String.valueOf(representation.isFetchRoles()));
+        }
         updateRoles(policy, authorization, representation.getRoles());
     }
 

--- a/core/src/main/java/org/keycloak/representations/idm/authorization/RolePolicyRepresentation.java
+++ b/core/src/main/java/org/keycloak/representations/idm/authorization/RolePolicyRepresentation.java
@@ -25,7 +25,7 @@ import java.util.Set;
 public class RolePolicyRepresentation extends AbstractPolicyRepresentation {
 
     private Set<RoleDefinition> roles;
-    private boolean fetchRoles;
+    private Boolean fetchRoles;
 
     @Override
     public String getType() {
@@ -40,7 +40,7 @@ public class RolePolicyRepresentation extends AbstractPolicyRepresentation {
         this.roles = roles;
     }
 
-    public void addRole(String name, boolean required) {
+    public void addRole(String name, Boolean required) {
         if (roles == null) {
             roles = new HashSet<>();
         }
@@ -59,24 +59,24 @@ public class RolePolicyRepresentation extends AbstractPolicyRepresentation {
         addRole(clientId + "/" + name, required);
     }
 
-    public boolean isFetchRoles() {
+    public Boolean isFetchRoles() {
         return fetchRoles;
     }
 
-    public void setFetchRoles(boolean fetchRoles) {
+    public void setFetchRoles(Boolean fetchRoles) {
         this.fetchRoles = fetchRoles;
     }
 
     public static class RoleDefinition implements Comparable<RoleDefinition> {
 
         private String id;
-        private boolean required;
+        private Boolean required;
 
         public RoleDefinition() {
             this(null, false);
         }
 
-        public RoleDefinition(String id, boolean required) {
+        public RoleDefinition(String id, Boolean required) {
             this.id = id;
             this.required = required;
         }
@@ -89,11 +89,11 @@ public class RolePolicyRepresentation extends AbstractPolicyRepresentation {
             this.id = id;
         }
 
-        public boolean isRequired() {
+        public Boolean isRequired() {
             return required;
         }
 
-        public void setRequired(boolean required) {
+        public void setRequired(Boolean required) {
             this.required = required;
         }
 

